### ISSUE TITLE
perf: AiDotNet-specific fused kernels — FusedLoRAForward / FusedSparseLinear / FusedDenoiseStep

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#301]
## Context

AiDotNet has several inference / training hot paths that currently decompose into 3–5 separate engine calls each, allocating per-stage intermediates and burning extra memory bandwidth. These shapes are unique enough to AiDotNet that adding them as first-class fused kernels in `AiDotNet.Tensors` (CPU + GPU) is the right call — they don't generalize as well-known ops PyTorch / TF would already provide, but they're the inner loops of the workloads people actually run.

This issue tracks three high-ROI fused-kernel candidates. Each is independently shippable; recommend taking them in the priority order below.

The pattern matches the existing fused-kernel infrastructure already in this repo:
- CPU fused ops live in [`src/AiDotNet.Tensors/Helpers/CpuFusedOperations.cs`](https://github.com/ooples/AiDotNet.Tensors/blob/main/src/AiDotNet.Tensors/Helpers/CpuFusedOperations.cs) (e.g. `FusedGemmBiasActivation`).
- GPU fused ops live under `src/AiDotNet.Tensors/Engines/DirectGpu/` (e.g. `FusedLinearKernels.cs` for OpenCL, the CUDA-side wrappers in `CudaBackend.*.cs`).
- Compile-time fusion is driven by [`CpuFusionPass`](https://github.com/ooples/AiDotNet.Tensors/blob/main/src/AiDotNet.Tensors/Engines/Compilation/CpuFusionPass.cs), which pattern-matches lazy graphs and emits fused `LazyNode<T>` instances.

Each new kernel ships:
1. A CPU implementation in `CpuFusedOperations`.
2. A CUDA implementation (via `CudaBackend` partial — match the layout of existing fused linear paths).
3. An OpenCL kernel string in `Engines/DirectGpu/OpenCL/Kernels/`.
4. A `LazyNodeType` entry + pattern matcher in `CpuFusionPass` so traced graphs use it automatically.
5. A benchmark in `tests/AiDotNet.Tensors.Benchmarks/` showing the win vs the decomposed path AND vs PyTorch/TorchSharp on the equivalent shape.
6. Numerical-equivalence test in `tests/AiDotNet.Tensors.Tests/Engines/Compilation/` against the eager-engine path.

Public surface stays minimal: each kernel is exposed via a static helper on `CpuFusedOperations` (or a sibling type for GPU-only kernels) — no additions to the `IEngine` interface (binary-breaking for external implementers per the long-standing #166 / #170 / #199 pattern).

---

## Workstream 1 — `FusedLoRAForward` (highest ROI; do first)

LoRA adapters are on every fine-tuned model in the AiDotNet ecosystem (LoRALayer, ChainLoRAAdapter, DeltaLoRAAdapter, GLoRAAdapter, NOLAAdapter, RoSAAdapter, TiedLoRAAdapter, X-LoRAAdapter, LoRETTAAdapter). The forward path is **the** inner loop of any LoRA-tuned inference call.

### Current decomposed path

```
hBase    = base.Forward(x)            // [batch, out]
intermed = MatMul(x, A)                // [batch, rank]   ← LIVE FOR ENTIRE STAGE
delta    = MatMul(intermed, B)        // [batch, out]
scaled   = MultiplyScalar(delta, α)   // [batch, out]
y        = Add(hBase, scaled)         // [batch, out]
```

5 engine calls. Two intermediate tensors (`intermed`, `delta`) materialized for the entire stage. Memory traffic: x is read twice, base output is read twice.

### Fused kernel signature

```csharp
public static void FusedLoRAForward(
    Tensor<float> input,        // [batch, in]
    Tensor<float> baseOutput,   // [batch, out] — pre-computed base.Forward(x)
    Tensor<float> loraA,        // [in, rank]
    Tensor<float> loraB,        // [rank, out]
    float scaling,              // alpha / rank
    Tensor<float> output);      // [batch, out] — written in place
```

Internally tiles by row of `input`: for each batch row, hold a `rank`-element accumulator in registers, do `acc = x_row · A` (rank-wide dot), then `y_row = baseOutput_row + scaling * (acc · B)` (out-wide accumulator). No materialized `[batch, rank]` intermediate; `rank` is typically 4–32 so the per-row state fits in registers.

For non-frozen base layers, the `+ baseOutput` term is just an in-place add to whatever the base layer wrote into `output` — saves the separate add.

### Expected wins (per call, BS=32 typical inference)

| Metric | Decomposed | Fused | Δ |
|---|---:|---:|---:|
| Engine calls | 5 | 1 | −80% |
| Intermediate alloc | ~`batch × rank × 4` + `batch × out × 4` | 0 | −100% |
| Memory traffic | 4× full pass | 1× | −75% |
| Latency (BS=32, in=out=2048, rank=8, est.) | ~95 µs | ~28 µs | **−70%** |

### Acceptance criteria

- CPU fused implementation in `CpuFusedOperations.FusedLoRAForward` (signature above).
- CUDA implementation reachable from `DirectGpuTensorEngine` for typical foundation-model shapes.
- New benchmark `LoRAForwardBenchmark.cs` comparing decomposed-engine vs fused-CPU vs fused-CUDA vs TorchSharp `nn.Linear + LoRA-equivalent` at BS ∈ {1, 32, 128} with rank ∈ {4, 16, 64}, in/out ∈ {1024, 2048, 4096}. Pass criterion: fused CPU ≥ 1.5× faster than decomposed at every shape; fused CUDA ≥ 5× faster than decomposed.
- Numerical-equivalence test against eager path within fp32 tolerance (1e-4 relative).
- `LazyNodeType.FusedLoRA` registered + `CpuFusionPass` detects the `MatMul(x, A) → MatMul(_, B) → MultiplyScalar(_, α) → Add(base, _)` pattern and emits the fused node automatically. Existing LoRA layers transparently get the win once this lands — no consumer-side code changes.
- Downstream wiring tracked separately under `ooples/AiDotNet#TBD` once the kernel is in place.

---

## Workstream 2 — `FusedDenoiseStep`

Diffusion sampling loops are 20–50 forward passes through a noise-predictor + a sampler-update step. The sampler update (DDPM / DDIM / Euler / DPMSolver++) is itself 5–10 elementwise ops. Today each step is *two* separate compiled plans (noise predictor + sampler) chained via `ChainAsync`, so per-step overhead = `chain boundary cost` + `5–10 elementwise alloc/dispatch overheads`.

For SDXL-class models the sampler-update overhead is small relative to the U-Net forward (the U-Net dominates), but for distilled models (LCM, Turbo, 1–4 step samplers) the U-Net is fast enough that the sampler-update overhead becomes a real fraction of total wall time.

### Variants to support

| Variant | Formula |
|---|---|
| DDPM | `x_{t-1} = (1/sqrt(α_t)) * (x_t − (β_t/sqrt(1-ᾱ_t)) * ε_θ) + σ_t * z` |
| DDIM | `x_{t-1} = sqrt(ᾱ_{t-1}) * x_0_pred + sqrt(1-ᾱ_{t-1}) * ε_θ` |
| Euler | `x_{t-1} = x_t + dt * ε_θ` |
| DPMSolver++ 2nd order | (2-stage; needs prev_noise as additional input) |

Ship DDIM first (most common for distilled / fast samplers), the others as additional fused-kernel variants in the same file.

### Fused kernel signature (DDIM)

```csharp
public static void FusedDDIMStep(
    Tensor<float> x_t,              // [batch, ...] — current latent
    Tensor<float> epsilon_theta,    // [batch, ...] — noise predictor output
    float alpha_bar_t,              // pre-computed schedule constant
    float alpha_bar_t_minus_1,      // pre-computed schedule constant
    Tensor<float> x_t_minus_1);     // [batch, ...] — written in place
```

Single elementwise pass. CPU: SIMD-vectorized over the flattened tensor. GPU: one kernel launch.

### Acceptance criteria

- CPU + CUDA fused DDIM step kernel.
- New benchmark `DiffusionSamplerBenchmark.cs` adding the fused variants alongside the existing `DiffusionPipelineBenchmark` from PR #298. Pass criterion: 1–4-step distilled sampler (the regime where sampler overhead matters) ≤ 1.10× of the decomposed path's CPU time, GPU same.
- Numerical-equivalence test against the eager / decomposed path.
- `LazyNodeType.FusedDDIMStep` + pattern matcher.
- Follow-on: DDPM, Euler, DPMSolver++ 2nd order.

---

## Workstream 3 — `FusedSparseLinear`

Sparse linear layers (RoSA, sparse fine-tuning) currently materialize the full `[out, in]` mask and run a dense matmul through it, which defeats the point of sparsity — the FLOPs are still O(in · out · batch) instead of O(nnz · batch). A real sparse-linear kernel reduces both FLOPs and memory traffic proportional to the sparsity ratio (typical 90–99%).

### Fused kernel signature

```csharp
public static void FusedSparseLinear(
    Tensor<float> input,            // [batch, in]
    int[] sparseRowOffsets,         // CSR row pointers, length out+1
    int[] sparseColIndices,         // CSR column indices, length nnz
    Tensor<float> sparseValues,     // [nnz]
    Tensor<float>? bias,            // [out] optional
    FusedActivationType activation, // None / ReLU / GELU / Sigmoid / Tanh
    Tensor<float> output);          // [batch, out]
```

CSR format. CPU: row-parallel outer loop (each output row sums `nnz_per_row` × batch products). GPU: warp-per-row + atomic accumulate, or CSR-vector pattern.

### Expected wins

| Sparsity | FLOPs reduction | Memory reduction |
|---:|---:|---:|
| 90% | 10× | 5×–10× |
| 95% | 20× | 10×–15× |
| 99% | 100× | 50×–80× |

### Acceptance criteria

- CPU + CUDA sparse-linear kernel.
- New benchmark `SparseLinearBenchmark.cs` at sparsity ∈ {90%, 95%, 99%} and shape ∈ {[B, 1024], [B, 4096]} comparing fused vs the existing dense-with-mask path. Pass criterion: at 95% sparsity, fused ≥ 5× faster than dense-with-mask on CPU, ≥ 10× faster on GPU.
- Numerical-equivalence test.
- RoSAAdapter wires through to the fused path automatically (no API change).

---

## Out of scope

- **Adding members to `IEngine`** — we keep the public engine surface frozen. New kernels are static helpers (CPU) or live on the concrete GPU backend type (GPU); the engine dispatches via `Engine.DirectGpu` for the GPU path.
- **Mixed-precision (fp16, bf16) kernels** — a separate epic; ship fp32 first, then add fp16 once the fused-kernel infrastructure has been validated.
- **Backward passes** for these kernels — inference-first; backward kernels for training-time use are tracked separately once forward lands.

## Why these and not generic ops

Standard ops (`MatMul`, `Conv2D`, `Softmax`, `LayerNorm`, etc.) already have well-tuned implementations in libtorch / MKL / cuBLAS that we wrap or replicate. The three above are AiDotNet-specific compositions: LoRA is a fine-tuning strategy not present in stock PyTorch as a single op; DDIM is a sampler that PyTorch ships only via Python-level loop bodies (no fused kernel); sparse linear in PyTorch is `torch.sparse` matmul which has known perf issues at 90%–99% sparsity. So beating PyTorch here is a question of "implement what they didn't" rather than "out-tune their hand-tuned kernel."

## Priority order

1. `FusedLoRAForward` (highest ROI, broadest applicability)
2. `FusedSparseLinear` (next-largest gap, RoSA + sparse fine-tuning audience)
3. `FusedDenoiseStep` (narrowest audience — distilled samplers — but cleanest implementation)

Spawn child issues for each workstream once an owner picks one up.
